pkg/streamingpromql/operators/selectors/extend_range_vector.go

```diff
@@ -100,45 +200,54 @@ func NewExtendedPointsForAnchored(view *types.FPointRingBufferView, rangeStart,
 //     the first point with T >= rangeEnd, or taken directly from
 //     the last point with T <= rangeEnd.
 //
-// When an interpolated boundary point is used, an additional interpolated
-// value—applying counter-reset compensation—is written into the provided
-// smoothedHead or smoothedTail. If the returned range is later used by
-// rate() or increase(), these alternate compensated boundary points must be
-// substituted for the default smoothed ones.
+// When calculateCounterAdjustedPoints is true, additional interpolated points
+// for smoothedHead and smoothedTail are returned.
 //
-// The returned slice includes all points within the specified range, along
-// with the added (interpolated or direct) boundary points.
+// These points include counter-reset compensation to align with the left and
+// right range boundaries. rate() and increase() use these adjusted boundary
+// points.
 //
 // This implementation is based on extendFloats() from promql/engine.go.
-func NewExtendedPointsForSmoothed(view *types.FPointRingBufferView, rangeStart, rangeEnd int64, memoryConsumptionTracker *limiter.MemoryConsumptionTracker) (SmoothedPoints, error) {
-	extendedPoints, err := NewExtendedPointsForAnchored(view, rangeStart, rangeEnd, memoryConsumptionTracker)
-	if err != nil {
-		return SmoothedPoints{}, err
-	}
+func ConvertExtendedPointsToSmoothed(anchoredMetadata AnchoredExtensionMetadata, buff *types.FPointRingBuffer, rangeStart, rangeEnd int64, calculateCounterAdjustedPoints bool) (SmoothedPoints, error) {
```

Comment:
I find it is not clear that we have `AnchoredExtensionMetadata` to track buffer changes for undo purpose which if I understand correctly despite the name prefixed with `Anchored` it is used for smoothed as well. Now we have `ConvertExtendedPointsToSmoothed` to get the interpolation for smoothed which returns `SmoothedPoints`. I am slightly wondering why there is no `AnchoredPoints` which maybe not necessary because it is handled directly in `AnchoredExtensionMetadata` itself. But I think it will be helpful if we can make the abstraction slightly clearer here.